### PR TITLE
AutoImputation

### LINKED ISSUE #1: Add hyperparameter tuning to each model
We should compare the best-hyperparameter version of each model for the comparisons to be fair.

### LINKED ISSUE #30: Publish package to PyPI

### LINKED ISSUE #33: AutoImpute
Args:
- Donor dataset [DataFrame]
- Receiver dataset [DataFrame]
- Predictor variables (should be the same names) [List]
- Imputed variables [List]
- Optional: Model list to test, hyperparameters, test/train split (or `k` in k-fold cv), quantiles to test, etc.

Automatically:
- For each method, run the validation routine for all imputed variables and save the quantile losses per quantile and imputed variable
- Define a total quantile loss
- Pick the best method
- Re-run imputation with full data using best method
- Return object containing test losses, method, **result**, etc.

### LINKED ISSUE #36: Apply statsmodels graphics interface
e.g. https://www.statsmodels.org/stable/generated/statsmodels.graphics.gofplots.qqplot.html#statsmodels.graphics.gofplots.qqplot

Question: should we move to matplotlib for consistency with statsmodels?

### LINKED ISSUE #41: Deploy documentation to GitHub pages
We just need to copy over this [bit](https://github.com/PolicyEngine/policyengine-uk/blob/master/.github/workflows/code_changes.yaml) from the UK action that builds and published the documentation, then enable GitHub page publishing in the repo settings.

### PR DESCRIPTION
**Pull request by juaristi22**
Fix #33 
Fix #41 
Fix #1
Fix #36 
Fix #30

### CODE REVIEW

File: microimpute/comparisons/autoimpute.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            log.error(error_msg)
+            raise ValueError(error_msg)
+
+        log.info(
+            f"Generating imputations to impute from {len(donor_data)} donor data to {len(receiver_data)} receiver data for variables {imputed_variables} with predictors {predictors}. "
+        )
+
+        # Step 1: Data preparation
+
+        ## Normalizing ?? should we drop the columns that are not predictors or imputed variables?
```
Comment (nikhilwoodruff) [inline microimpute/comparisons/autoimpute.py:81]: So I think we want to preserve as much of the input data as possible, so let's make a copy of the passed dataframes at the start and assign over them at the end.

Comment (nikhilwoodruff) [reply microimpute/comparisons/autoimpute.py:81]: And let's make sure to reverse any normalisation or processing at the end so they get back exactly what they're asking for, which is the imputations in the same format as the data they passed in.

File: microimpute/comparisons/autoimpute.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        Y_test = X_test[imputed_variables]
+
+        # If imputed variables are in receiver data, remove them
+        receiver_data = receiver_data.drop(
+            columns=imputed_variables, errors="ignore"
+        )
+
+        # Step 2: Imputation with each method
+
+        ## We evaluate imputation methods with a test donor subset right? And we train on all passed data and hope the user makes decisions about size?
```
Comment (nikhilwoodruff) [inline microimpute/comparisons/autoimpute.py:114]: Correct!

File: microimpute/comparisons/autoimpute.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+        if not models:
+            # If no models are provided, use default models
+            model_classes: List[Type[Imputer]] = [QRF, OLS, QuantReg, Matching]
+
+        ## How do we want to handle the hyperparameters? Do we want to optimize in autoimpute as well and restructure all other functions to accept hyperparameters?
```
Comment (nikhilwoodruff) [inline microimpute/comparisons/autoimpute.py:120]: Totally OK to separate the hyperparameter tuning out from this PR and file later- it's not critical that it's in here, though we do need to add it at some point.

Comment (nikhilwoodruff) [reply microimpute/comparisons/autoimpute.py:120]: But yes tuning should go in this PR.

File: microimpute/comparisons/autoimpute.py (hunk 1 above)
Comment (juaristi22) [reply microimpute/comparisons/autoimpute.py]: The imputation is complete, and there is unnormalization at the end of the process. I just wasn't sure how we integrate it back into the dataset, do we chose a single quantile (maybe a parameter for the user to specify?) or how do we choose which quantile best predicts each datapoint?

File: microimpute/evaluations/cross_validation.py [hunk 4]
```diff
@@ -110,19 +117,56 @@ def cross_validate_model(
...
                 # Handle different model fitting requirements
-                if model_class == QuantReg:
-                    log.info(f"Fitting QuantReg model with explicit quantiles")
-                    fitted_model = model.fit(
-                        train_data,
-                        predictors,
-                        imputed_variables,
-                        quantiles=quantiles,
-                    )
+                if model_hyperparams and model_class.__name__ == "QRF":
```
Comment (juaristi22) [inline microimpute/evaluations/cross_validation.py:120]: currently hyperparameter setting and tuning are only enabled for QRF, would love to discuss if it is possible / worth it to implement for others (eg. we could try tuning for matching with and without replacement but I'm not sure if it is worth it computationally given that matching seems to perform relatively worse anyway)

File: microimpute/models/qrf.py [hunk 5]
```diff
@@ -158,3 +175,97 @@ def _fit(
...
+        optuna.logging.set_verbosity(optuna.logging.WARNING)
+
+        # Create a validation split (80% train, 20% validation)
+        X_train, X_test = train_test_split(
+            data, test_size=0.2, random_state=self.seed
+        )
+
+        def objective(trial: optuna.Trial) -> float:
+            # we may want to add / change some hyperparameters
+            params = {
```
Comment (juaristi22) [inline microimpute/models/qrf.py:209]: atm we only tune these 5 hyperparameters which include the ones identified as most important in most literature (maybe with the exception of max_depth?) should we decide otherwise on these?

File: docs/autoimpute/autoimpute.ipynb [hunk 6]
```diff
@@ -0,0 +1,4943 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/Users/movil1/anaconda3/lib/python3.11/site-packages/statsmodels/regression/quantile_regression.py:191: IterationLimitWarning: Maximum number of iterations (1000) reached.\n",
```
Comment (juaristi22) [inline docs/autoimpute/autoimpute.ipynb:237]: I can't manage to supress the warnings showing in the documentation :((

File: microimpute/evaluations/cross_validation.py (hunk 4 above)
Comment (juaristi22) [reply microimpute/evaluations/cross_validation.py:120]: and, hyperparameter setting and tuning at the same time raises an error asking the user to choose one or the other. maybe we want to choose which one should override the other?

Comment (nikhilwoodruff) [reply microimpute/evaluations/cross_validation.py:120]: > sure if it is worth it computationally given that matching seems to perform relatively worse anyway

I think the reason would be to make sure it's a completely fair comparison.

> and, hyperparameter setting and tuning at the same time raises an error asking the user to choose one or the other. maybe we want to choose which one should override the other?

It's totally fine to raise the error, we can be forceful about avoiding user error.

File: .github/workflows/main.yml
Comment (nikhilwoodruff) [inline .github/workflows/main.yml:1]: Nice, thanks

File: docs/_config.yml [hunk 7]
```diff
@@ -12,6 +13,7 @@ repository:
 
 sphinx:
   config:
+    suppress_warnings: [""]
```
Comment (nikhilwoodruff) [inline docs/_config.yml:16]: If this doesn't do anything as you're finding, would remove

File: docs/_config.yml [hunk 8]
```diff
@@ -20,8 +22,8 @@ sphinx:
       - style.css
   extra_extensions:
     - "sphinx.ext.autodoc"
-    - "sphinxarg.ext"
+    #- "sphinxarg.ext"
```
Comment (nikhilwoodruff) [inline docs/_config.yml:25]: Delete rather than comment

File: docs/autoimpute/autoimpute.ipynb (hunk 6 above)
Comment (juaristi22) [reply docs/autoimpute/autoimpute.ipynb]: ah yes! when putting in the code cell rather than at the top of the notebook, thanks!

File: microimpute/models/quantreg.py [hunk 9]
```diff
@@ -179,9 +181,10 @@ def _fit(
                         ).fit(q=q)
                     self.logger.info(f"Model for q={q} fitted successfully")
             else:
-                q = np.random.uniform(0, 1)
+                random_generator = np.random.default_rng(self.seed)
```
Comment (juaristi22) [inline microimpute/models/quantreg.py:184]: Quantreg fits the model to a specific quantile, we can take this as the model being naturally constrained to predicting on the same quantile for all data points (which can be a random quantile, default to 0.5, or be the one that showed lowest loss in the comparison step). Or we can go through the hustle of fitting a set of (x models eg.10 one per quantile), and then for every data point in the dataset sample from the fitted quantiles, retrieve the imputations on the whole dataset for that quantile, and take the value that corresponds to the specific data point.

File: microimpute/models/ols.py [hunk 10]
```diff
@@ -102,16 +116,33 @@ def _predict_quantile(
...
+            normal_quantiles = norm.ppf(beta_samples)
+
+            # For each mean prediction, randomly select one of the quantiles
+            sampled_indices = random_generator.integers(
+                0, count_samples, size=len(mean_preds)
+            )
+            selected_quantiles = normal_quantiles[sampled_indices]
+
+            # Adjust each mean prediction by corresponding sampled quantile times standard error
+            return mean_preds + selected_quantiles * se
```
Comment (juaristi22) [inline microimpute/models/ols.py:137]: Following from the implementation in QRF, OLS adjusts the mean imputation relative to a specific quantile, which is randomly sampled from a beta distribution with alpha `a` (which depends on the mean_quantile defaulting to 0.5). I maintained count_samples=10. Do we like this? Because if we choose to do the whole fitting-all-data-to-every-quantile-then-sampling in QuantReg, we could also do the same for OLS, but I think that would introduce a bit more variation around the quantiles. The difference is probably not that big anyway?

**Review by nikhilwoodruff — APPROVED** [review]
